How many late days are you using for this assignment? 0 late days.

In [35]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

# Part 1: Load the dataset

In [6]:
# Load the given datasets
categorical = pd.read_csv('chronic_kidney_disease_categorical.csv')
numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
print(categorical)
print(numerical)

     unique_id   al   su       rbc        pc  ...  cad appet   pe  ane  Target
0       224481  3.0  NaN       NaN       NaN  ...   no  good  yes   no     ckd
1       992643  0.0  0.0       NaN       NaN  ...   no  good   no   no  notckd
2       308740  1.0  0.0  abnormal  abnormal  ...   no  good  yes   no     ckd
3       450314  1.0  0.0       NaN    normal  ...   no  good   no  yes     ckd
4       881763  0.0  0.0    normal    normal  ...   no  good   no   no  notckd
..         ...  ...  ...       ...       ...  ...  ...   ...  ...  ...     ...
425     769512  0.0  0.0    normal    normal  ...   no  good   no   no  notckd
426     872034  2.0  1.0  abnormal  abnormal  ...  yes  poor   no   no     ckd
427     643007  NaN  NaN       NaN    normal  ...  yes  poor   no   no     ckd
428     840392  NaN  NaN       NaN       NaN  ...   no  poor   no   no     ckd
429     856244  0.0  0.0       NaN    normal  ...   no  poor   no   no     ckd

[430 rows x 14 columns]
     unique_id   age     bp

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data has either numerical and categorical features (not both) that can be described as integers, binary, or continuous that provide important medical data about patients. Understanding these features and their types is crucial for effective data analysis and modeling. Here is some of what the labels are and what they represent.

- Numerical Features examples:
- `age`: A numerical value representing the age of the patient.
- `bp`: Blood pressure, which is typically a numerical measurement (e.g., systolic/diastolic).
- `bgr`: Blood glucose random, a numerical value indicating blood glucose levels.
- Categorical Features examples:
- `al`: Albumin level, which might indicate if it's normal or abnormal (categorical).
- `su`: Sugar in urine, which can indicate presence (yes/no).
- `htn`: Hypertension (yes/no).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [7]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = numerical.duplicated(subset=['unique_id'], keep='last').sum()
if numerical_duplicates > 0:
    print("Duplicates rows in numerical dataset based on 'unique_id':")
    print(numerical_duplicates)

# Drop these duplicate rows
numerical_cleaned = numerical.drop_duplicates(subset=['unique_id'], keep='last')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = categorical.duplicated(subset=['unique_id'], keep='last').sum()
if categorical_duplicates > 0:
    print("Duplicates rows in numerical dataset based on 'unique_id':")
    print(categorical_duplicates)
categorical_cleaned = categorical.drop_duplicates(subset=['unique_id'], keep='last')


Duplicates rows in numerical dataset based on 'unique_id':
30
Duplicates rows in numerical dataset based on 'unique_id':
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique IDs are present in both datasets. I think this is the case because some IDs only have numerical or categorical data only but not both because they might be from different clinics or maybe one side of the ID got deleted in one dataset because it was outdated. When an ID/row is missing, you see rows with NaNs or blank places for columns that belong to the dataset where the ID is not present.

In [8]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_dataset = pd.merge(numerical_cleaned, categorical_cleaned, on="unique_id", how="outer")
#Print the combined dataset
print(merged_dataset)

     unique_id   age    bp    bgr     bu  ...  cad  appet   pe  ane  Target
0       102677  59.0  70.0   76.0  186.0  ...   no   poor  yes  yes     ckd
1       103570  24.0  80.0  125.0    NaN  ...   no   good   no   no  notckd
2       104675  65.0  80.0  215.0  133.0  ...   no   good   no   no     ckd
3       105923  90.0  90.0  139.0   89.0  ...   no   good   no   no     ckd
4       107297  60.0  70.0    NaN    NaN  ...   no   good   no   no  notckd
..         ...   ...   ...    ...    ...  ...  ...    ...  ...  ...     ...
395     995177  69.0  70.0  214.0   96.0  ...  yes   good  yes  yes     ckd
396     996076  57.0  60.0  105.0   49.0  ...   no   good   no   no  notckd
397     996412  70.0  60.0  120.0   40.0  ...   no   good   no   no  notckd
398     997629  35.0   NaN   99.0   30.0  ...   no   good   no   no  notckd
399     997876  64.0  70.0  107.0   15.0  ...   no   good   no   no     ckd

[400 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [9]:
# Calculate the percentage of rows that contain at least one missing value
total_rows = len(merged_dataset)
missing_value_rows = merged_dataset.isnull().any(axis=1).sum()
missing_percentage = (missing_value_rows / total_rows) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {missing_percentage:.2f}%")
# Drop these rows from the dataset
cleaned_dataset = merged_dataset.dropna()
# Print the Dataset
print(cleaned_dataset)

Percentage of rows with at least one missing value: 61.75%
     unique_id   age    bp    bgr     bu  ...  cad  appet   pe  ane  Target
0       102677  59.0  70.0   76.0  186.0  ...   no   poor  yes  yes     ckd
6       109053  64.0  60.0  106.0   27.0  ...   no   good   no   no  notckd
7       114717  41.0  80.0  122.0   25.0  ...   no   good   no   no  notckd
8       118191  42.0  70.0   93.0   32.0  ...   no   good   no   no  notckd
9       118596  74.0  60.0   88.0   50.0  ...   no   good   no   no  notckd
..         ...   ...   ...    ...    ...  ...  ...    ...  ...  ...     ...
390     980291  62.0  70.0  122.0   42.0  ...   no   good   no   no     ckd
391     991031  30.0  80.0   89.0   42.0  ...   no   good   no   no  notckd
394     994377  71.0  70.0   81.0   18.0  ...   no   good   no   no  notckd
395     995177  69.0  70.0  214.0   96.0  ...  yes   good  yes  yes     ckd
396     996076  57.0  60.0  105.0   49.0  ...   no   good   no   no  notckd

[153 rows x 25 columns]


## Part 3.4: Sort the dataset according to the Labels

In [10]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_dataset_sorted = cleaned_dataset.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(merged_dataset_sorted)

     unique_id   age    bp    bgr     bu  ...  cad  appet   pe  ane  Target
0       102677  59.0  70.0   76.0  186.0  ...   no   poor  yes  yes     ckd
1       337834  48.0  70.0  117.0   56.0  ...   no   poor  yes  yes     ckd
2       343710  61.0  80.0  173.0  148.0  ...  yes   poor  yes  yes     ckd
3       349892  59.0  70.0  424.0   55.0  ...  yes   good   no   no     ckd
4       353025  40.0  70.0  253.0  150.0  ...   no   poor  yes   no     ckd
..         ...   ...   ...    ...    ...  ...  ...    ...  ...  ...     ...
148     486397  20.0  70.0  123.0   44.0  ...   no   good   no   no  notckd
149     203466  22.0  60.0   97.0   18.0  ...   no   good   no   no  notckd
150     482105  44.0  60.0   95.0   46.0  ...   no   good   no   no  notckd
151     228849  58.0  80.0  100.0   50.0  ...   no   good   no   no  notckd
152     996076  57.0  60.0  105.0   49.0  ...   no   good   no   no  notckd

[153 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [11]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dataset_dummy = pd.get_dummies(merged_dataset_sorted, dtype=int, columns=categorical_columns, drop_first=True)
# Print the dataset
print(dataset_dummy)

     unique_id   age    bp    bgr  ...  appet_poor  pe_yes  ane_yes  Target_notckd
0       102677  59.0  70.0   76.0  ...           1       1        1              0
1       337834  48.0  70.0  117.0  ...           1       1        1              0
2       343710  61.0  80.0  173.0  ...           1       1        1              0
3       349892  59.0  70.0  424.0  ...           0       0        0              0
4       353025  40.0  70.0  253.0  ...           1       1        0              0
..         ...   ...   ...    ...  ...         ...     ...      ...            ...
148     486397  20.0  70.0  123.0  ...           0       0        0              1
149     203466  22.0  60.0   97.0  ...           0       0        0              1
150     482105  44.0  60.0   95.0  ...           0       0        0              1
151     228849  58.0  80.0  100.0  ...           0       0        0              1
152     996076  57.0  60.0  105.0  ...           0       0        0              1

[15

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Ultimately, the choice between the two approaches depends on the specific context of your data, the machine learning models you plan to use, and the nature of the categorical variable. For binary variables, both methods are generally equivalent. The effect of choosing a single binary column for categorical variables with more than two values can reduce the complexity of the model making it easier to understand, simplifies any data processing, but limits the model flexibility.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [12]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = dataset_dummy[column].mean()
    std = dataset_dummy[column].std()
    lower_bound = mean - 3 * std
    upper_bound = mean + 3 * std
    dataset_dummy = dataset_dummy[(dataset_dummy[column] >= lower_bound) & (dataset_dummy[column] <= upper_bound)]
# Print the dataset
print(dataset_dummy)

     unique_id   age     bp    bgr  ...  appet_poor  pe_yes  ane_yes  Target_notckd
2       343710  61.0   80.0  173.0  ...           1       1        1              0
5       397388  63.0  100.0  280.0  ...           0       0        0              0
7       438182  73.0   80.0  253.0  ...           0       0        0              0
8       474407  21.0   90.0  107.0  ...           0       0        1              0
9       475200  64.0  100.0  163.0  ...           0       1        0              0
..         ...   ...    ...    ...  ...         ...     ...      ...            ...
148     486397  20.0   70.0  123.0  ...           0       0        0              1
149     203466  22.0   60.0   97.0  ...           0       0        0              1
150     482105  44.0   60.0   95.0  ...           0       0        0              1
151     228849  58.0   80.0  100.0  ...           0       0        0              1
152     996076  57.0   60.0  105.0  ...           0       0        0        

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [13]:
# Normalize the all Numerical Attributes in the dataset.
scaler = StandardScaler()
dataset_dummy[numerical_columns] = scaler.fit_transform(dataset_dummy[numerical_columns])
# Print the dataset
print(dataset_dummy)

     unique_id       age        bp  ...  pe_yes  ane_yes  Target_notckd
2       343710  0.810190  0.730553  ...       1        1              0
5       397388  0.936367  2.776102  ...       0        0              0
7       438182  1.567253  0.730553  ...       0        0              0
8       474407 -1.713353  1.753327  ...       0        1              0
9       475200  0.999456  2.776102  ...       1        0              0
..         ...       ...       ...  ...     ...      ...            ...
148     486397 -1.776441 -0.292221  ...       0        0              1
149     203466 -1.650264 -1.314996  ...       0        0              1
150     482105 -0.262316 -1.314996  ...       0        0              1
151     228849  0.620924  0.730553  ...       0        0              1
152     996076  0.557836 -1.314996  ...       0        0              1

[133 rows x 25 columns]


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column is not appropriate for modeling and predictions because it is not related to kidney disease since its values give no information on it. It is only included to help organize the data. 

In [14]:
#Remove that column
dataset_dummy = dataset_dummy.drop(columns={'unique_id'})
# Print the dataset
print(dataset_dummy)

          age        bp       bgr  ...  pe_yes  ane_yes  Target_notckd
2    0.810190  0.730553  1.146438  ...       1        1              0
5    0.936367  2.776102  3.520558  ...       0        0              0
7    1.567253  0.730553  2.921481  ...       0        0              0
8   -1.713353  1.753327 -0.317973  ...       0        1              0
9    0.999456  2.776102  0.924557  ...       1        0              0
..        ...       ...       ...  ...     ...      ...            ...
148 -1.776441 -0.292221  0.037036  ...       0        0              1
149 -1.650264 -1.314996 -0.539853  ...       0        0              1
150 -0.262316 -1.314996 -0.584229  ...       0        0              1
151  0.620924  0.730553 -0.473289  ...       0        0              1
152  0.557836 -1.314996 -0.362349  ...       0        0              1

[133 rows x 24 columns]


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [19]:
# Export the dataframe to a new csv file
dataset_dummy.to_csv('cleaned_dataframe.csv', index=False)
# Export the dataframe to a new json file
dataset_dummy.to_json('cleaned_dataframe.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [ ]:
print(dataset_dummy.head(15))

| | age | bp | bgr | bu | sc | sod | pot | hemo |
|----|------------|-----------|------------|-----------|-----------|------------|------------|-----------|
| 2 | 0.810190 | 0.730553 | 1.146438 | 4.014305 | 1.690051 | -0.959702 | 1.373024 | -2.913221 |
| 5 | 0.936367 | 2.776102 | 3.520558 | -0.225995 | 1.218114 | 0.448743 | -1.303189 | -0.607180 |
| 7 | 1.567253 | 0.730553 | 2.921481 | 3.789156 | 2.161987 | -0.431535 | 2.317570 | -1.694935 |
| 8 | -1.713353 | 1.753327 | -0.317973 | -0.038371 | 0.206821 | -2.720258 | -1.303189 | -2.652160 |
| 9 | 0.999456 | 2.776102 | 0.924557 | 0.486976 | 3.914895 | -0.079424 | 0.428478 | -2.826200 |
| 10 | 0.747101 | -1.314996 | 3.698062 | -0.188470 | 0.206821 | -1.839980 | -2.090311 | -2.826200 |
| 12 | 0.494747 | 1.753327 | 0.170164 | 2.475789 | 3.577798 | -1.663925 | 0.743327 | -2.304078 |
| 13 | 1.314898 | -0.292221 | 2.056147 | 2.063016 | 3.308119 | -3.600537 | -0.673492 | -2.173547 |
| 15 | -0.136138 | -1.314996 | 0.924557 | 1.912917 | 1.285533 | 0.096632 | -0.516068 | -1.999507 |
| 18 | 0.873279 | -0.292221 | 0.014848 | 0.036678 | 0.206821 | -0.783646 | 0.585903 | -0.781221 |
| 19 | 0.431659 | 0.730553 | 2.056147 | 1.199946 | 1.690051 | -0.607591 | 0.900751 | -1.520894 |
| 20 | 0.999456 | -1.314996 | 2.610848 | 0.637075 | 1.959729 | -0.607591 | 1.687873 | -2.130037 |
| 22 | 1.441076 | -1.314996 | -0.073905 | 3.151235 | 2.633924 | -0.783646 | 0.900751 | -1.303343 |
| 23 | 1.441076 | -0.292221 | 2.167088 | 1.537669 | 1.487792 | -1.311813 | 0.113630 | -1.738445 |
| 29 | 0.684013 | 0.730553 | 4.030883 | -0.225995 | -0.062857 | -3.248425 | -1.303189 | -1.738445 |
| | pcv | wbcc | rbcc | al | su | rbc_normal | pc_normal | pcc_present |
|----|-----------|-----------|-----------|------|------|---------------|--------------|---------------|
| 2 | -2.905503 | 0.519658 | -2.301962 | 2.0 | 0.0 | 0 | 0 | 0 |
| 5 | -0.590673 | 0.785483 | -1.084138 | 2.0 | 2.0 | 1 | 1 | 0 |
| 7 | -1.603411 | -0.366426 | -0.962356 | 2.0 | 0.0 | 0 | 0 | 0 |
| 8 | -3.050180 | 1.937392 | -1.449485 | 4.0 | 0.0 | 1 | 0 | 1 |
| 9 | -2.616150 | -0.233513 | -2.058397 | 4.0 | 2.0 | 0 | 0 | 0 |
| 10 | -2.760826 | 3.177909 | -2.545527 | 3.0 | 1.0 | 1 | 0 | 1 |
| 12 | -2.182119 | -0.720859 | -2.058397 | 2.0 | 0.0 | 0 | 0 | 0 |
| 13 | -2.326796 | 1.538654 | -2.180180 | 4.0 | 3.0 | 1 | 0 | 1 |
| 15 | -2.326796 | 2.912084 | -2.301962 | 1.0 | 0.0 | 1 | 1 | 0 |
| 18 | -0.735350 | -0.056296 | -1.449485 | 3.0 | 0.0 | 1 | 0 | 0 |
| 19 | -1.458735 | -0.277817 | -1.693050 | 3.0 | 1.0 | 1 | 0 | 1 |
| 20 | -2.182119 | -0.233513 | -2.058397 | 4.0 | 1.0 | 0 | 0 | 0 |
| 22 | -1.314058 | 3.177909 | -0.962356 | 4.0 | 0.0 | 1 | 1 | 0 |
| 23 | -1.603411 | -1.075292 | -1.814833 | 3.0 | 0.0 | 1 | 0 | 1 |
| 29 | -1.314058 | 1.272829 | -0.962356 | 1.0 | 0.0 | 0 | 1 | 0 |
| | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_notckd |
|----|------------|---------|--------|---------|------------|--------|---------|----------------|
| 2 | 0 | 1 | 1 | 1 | 1 | 1 | 1 | 0 |
| 5 | 1 | 1 | 0 | 1 | 0 | 0 | 0 | 0 |
| 7 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 0 |
| 8 | 1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 |
| 9 | 1 | 1 | 1 | 0 | 0 | 1 | 0 | 0 |
| 10 | 0 | 1 | 0 | 0 | 1 | 0 | 0 | 0 |
| 12 | 0 | 1 | 0 | 0 | 0 | 0 | 0 | 0 |
| 13 | 1 | 1 | 1 | 1 | 0 | 1 | 0 | 0 |
| 15 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 0 |
| 18 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 0 |
| 19 | 1 | 1 | 1 | 0 | 0 | 1 | 0 | 1 |
| 20 | 1 | 1 | 1 | 0 | 1 | 1 | 0 | 0 |
| 22 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 0 |
| 23 | 0 | 1 | 1 | 1 | 0 | 0 | 0 | 0 |
| 29 | 0 | 0 | 1 | 0 | 1 | 0 | 0 | 0 |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

An effective way to mitigate data conversion hallucinations from LLM conversions would be to just simply test on smaller subsets of data points, to make sure the conversion works properly.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
# Code to rename all the columns in the dataset
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns of the dataframe
dataset_dummy.rename(columns=column_mapping, inplace=True)
# Displaying updated dataframe
print(dataset_dummy)

          age  blood pressure  blood glucose random  blood urea  \
2    0.810190        0.730553              1.146438    4.014305   
5    0.936367        2.776102              3.520558   -0.225995   
7    1.567253        0.730553              2.921481    3.789156   
8   -1.713353        1.753327             -0.317973   -0.038371   
9    0.999456        2.776102              0.924557    0.486976   
10   0.747101       -1.314996              3.698062   -0.188470   
12   0.494747        1.753327              0.170164    2.475789   
13   1.314898       -0.292221              2.056147    2.063016   
15  -0.136138       -1.314996              0.924557    1.912917   
18   0.873279       -0.292221              0.014848    0.036678   
19   0.431659        0.730553              2.056147    1.199946   
20   0.999456       -1.314996              2.610848    0.637075   
22   1.441076       -1.314996             -0.073905    3.151235   
23   1.441076       -0.292221              2.167088    1.53766

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [34]:
# Converted SQL to Pandas code
target_count = dataset_dummy.groupby('Target_notckd').size().reset_index(name='count')
# Displaying the result
print(target_count)

   Target_notckd  count
0              0     22
1              1    111


*Describe what the above code does here* This query counts how many times each value of Target does not appear in the dataset.